scrapper.py

```diff
@@ -21,43 +40,160 @@ class IncorrectNumberOfArticlesError(Exception):
     """
 
 
+def _clean_text(text):
+    return re.sub(r"[\n\t ]+", " ", text).strip()
+
+
+def _get_page(link):
+    print("attempting to connect to %s", link)
+    #time.sleep(random.uniform(2.0, 4.0))
+    headers = Headers().generate()
+    print(headers)
+    response = requests.get(link,
+                            headers=headers,
+                            timeout=5)
+    if not response.ok:
+        return None
+    return BeautifulSoup(response.text, "html.parser")
+    # except requests.exceptions.ConnectionError:
+    #    print("failed to connect to %s. trying again", link)
+        # return _get_page(link)
+    #    raise ServerThrottledError
+    #except requests.exceptions.ReadTimeout:
+        #print("timed out connecting to %s. trying again", link)
+        #return _get_page(link)
+
+
 class Crawler:
     """
     Crawler implementation
     """
+
     def __init__(self, seed_urls, max_articles: int):
-        pass
+        self.seed_urls = seed_urls
+        self.max_articles = max_articles
+        self.urls = []
 
     def _extract_url(self, article_bs):
-        pass
+        if not article_bs:
+            return
+        for node in article_bs.find_all("a", {"class": "file"}):
+            # ignore links to issues, only collect articles.
+            # this does not leave out any content, because issues
+            # are comprised of the same articles.
+            if "issue" in node["href"]:
+                continue
+            if len(self.urls) == self.max_articles:
+                break
+            self.urls.append(node["href"])
 
     def find_articles(self):
         """
         Finds articles
         """
-        pass
+        for seed in self.seed_urls:
+            if len(self.urls) == self.max_articles:
+                break
+            try:
+                self._extract_url(_get_page(seed))
+            except requests.exceptions.ConnectionError:
+                print("failed to connect to", seed)
+                continue
 
     def get_search_urls(self):
         """
         Returns seed_urls param
         """
-        pass
+        return self.seed_urls
+
+
+class HTMLParser:
+    def __init__(self, article_url, article_id):
+        self._pdf_url = article_url.replace("view", "download") + ".pdf"
+        self.article_url = "/".join(article_url.split("/")[:-1])
+        self.article_id = article_id
+        self.article = Article(url=article_url, article_id=article_id)
+
+    def parse(self):
+        self._fill_article_with_text()
+        article_bs = _get_page(self.article_url)
+        self._fill_article_with_meta_information(article_bs)
+        return self.article
+
+    def _fill_article_with_text(self):
+        #time.sleep(random.uniform(0.0, 1.0))
+        pdf_raw = PDFRawFile(self._pdf_url, self.article_id)
+        pdf_raw.download()
+        self.article.text = pdf_raw.get_text()
```

Comment:
there should be no references in the final text

Reply:
Could you elaborate what this means?